Clarify that jnp.clip gives preference to maximum values where bounds are incongruent

johannahaffner opened the pull request:
This is what `jnp.clip` already does, I think it would be good to add this to its documentation. 

```python
jnp.clip(jnp.array([3., 2., 4.]), min=jnp.array([4., 4., 4.]), max=jnp.array([2., 2., 2.]))
# Array([2., 2., 2.])
```

johannahaffner committed:
Clarify that jnp.clip gives preference to maximum values where interval bounds are incongruent

This is what `jnp.clip` already does, I think it would be good to add this to its documentation. 

```python
jnp.clip(jnp.array([3., 2., 4.]), min=jnp.array([4., 4., 4.]), max=jnp.array([2., 2., 2.]))
#Array([2., 2., 2.])
```

johannahaffner commented:
Good point! Happy to add the tests, if you can let me know where this should be added.

johannahaffner commented:
Roundabout here, perhaps? 

https://github.com/jax-ml/jax/blob/0cedff9f5d65c80f59e831e7a596493807a7ab27/tests/lax_numpy_test.py#L1024

psanal35 committed:
Skip CSR matmat/matvec float tests on ROCm <6.4 (NaN issue with beta==0).

Cherry-picked from ROCm fork (commit 5dbfa9d1bcb3629658c2ec9addf45ac389f17305).

Added TODOs to remove this check when ROCm 6.4+ is the minimum supported version.

jakevdp commented:
> Roundabout here, perhaps?

I don't think any of the test cases there have `a_max < a_min`

mattjj committed:
[hijax] landing prototype pieces

shouldn't affect existing behaviors, or trace time

The main implementation ideas:
* each Trace is tagged with a `requires_low: bool`
* each Jaxpr
  * is tagged with an `is_high: bool`, default False but set True while tracing
    if any hijax primitives are encountered
  * includes an `mut_types: dict[Var, HijaxType]` indicating final types for
    type-changing mutable hijax types
* each AbstractValue is tagged by a `mutable: bool` which is read to populate
  `mut_types`
* each Primitive
  * has an `is_high(**params) -> bool` method (depends on params for HOPs)
  * has a `to_lojax(*args, **params)` method taking and returning
    hijaxtypes-wrapping-lowtracers
* in `Primitive.bind`, we check if `prim.is_high(**params) and
  trace.requires_low`, and if so we call `prim.to_lojax`

Co-authored-by: Dougal Maclaurin <dougalm@google.com>

jkr26 committed:
Global find/replace in jax for s/divisble/divisible.

Spotted this in an error message, then saw there were a few other places with this typo.

PiperOrigin-RevId: 758416484

pschuh committed:
Prepare to make DmaCopyChunk movable.

PiperOrigin-RevId: 758436699

Commit:
allow accelerated deprecations from jax.tree_util

PiperOrigin-RevId: 758500294

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/9c00cd639a2076210ecc68e381ad242b95d46be4.

PiperOrigin-RevId: 758618886

hawkinsp committed:
Simplify if_building_jaxlib macro.

I don't believe the GPU case of this macro ever matters, so this can be a condition strictly about how we're building jaxlib. No behavioral changes intended.

PiperOrigin-RevId: 758624138

vfdev-5 committed:
Removed few tsan cpython suppressions as fixed

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/c7dbe1e216d132e7b4042907b8ba9454535edc70.

PiperOrigin-RevId: 758642769

bchetioui committed:
[Pallas/Mosaic GPU] Optimize the construction of output buffers for `plgpu.kernel`.

Previously to this change, we would zero initialize all buffers---which was a
big overhead. Instead, we now generate an empty custom call in order to only
generate an allocation.

PiperOrigin-RevId: 758659576

superbobry committed:
[pallas:mosaic_gpu] Do not do unnecessary `commit_smem_to_gmem_group` in `emit_pipeline`

If the loop does no copies, there is nothing to commit.

PiperOrigin-RevId: 758660634

bchetioui committed:
[Pallas/Mosaic GPU] Allow `_is_known_divisible` to always return `True` when the divisor is `1`.

PiperOrigin-RevId: 758670548

nitins17 committed:
Replace gsutil command with gcloud storage commands.

GCP recommends using `gcloud storage` instead of `gsutil`

https://cloud.google.com/storage/docs/gsutil

PiperOrigin-RevId: 758672654

yashk2810 committed:
Rename `SdyArraySharding -> SdyArray` and `SdyDimSharding -> SdyDim` since these are not `Sharding` from a JAX POV.

PiperOrigin-RevId: 758679923

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/caef87114bc75bd89f56b8562d32cd7b8887319f.

PiperOrigin-RevId: 758707765

majnemer committed:
[ifrt] Refactor away from deprecated constructors

PiperOrigin-RevId: 758715355

jakevdp committed:
support accelerated deprecation for jax.extend.ffi

yashk2810 committed:
Lower unreduced to shardy. GSPMD doesn't support unreduced.

PiperOrigin-RevId: 758719692

Google-ML-Automation committed:
Merge pull request #28635 from ROCm:skip-csr-mat-tests-rocm63

PiperOrigin-RevId: 758726482

Google-ML-Automation committed:
Merge pull request #28743 from jakevdp:ffi-test

PiperOrigin-RevId: 758726503

Google-ML-Automation committed:
Merge pull request #28732 from vfdev-5:remove-tsan-suppresions

PiperOrigin-RevId: 758726514

yashk2810 committed:
Make device_put work for python scalars when the sharding is not fully addressable i.e. `device_put(1, global_sharding)`. This already works for numpy arrays. Fixes: https://github.com/jax-ml/jax/discussions/14578#discussioncomment-13145332

PiperOrigin-RevId: 758730649

yashk2810 committed:
Partial in `repeats` if `total_repeat_length is None` since repeats is expected to be a constant.

This fixes an error in explicit sharding mode where we were converting repeats to a Tracer instead of it being a concrete value.

PiperOrigin-RevId: 758743044

dfm committed:
Add a linearization rule for scan.

jakevdp committed:
Warn when __jax_array__ is seen in xla dtype canonicalization

hawkinsp committed:
Add requirements needed for building wheels under Python 3.14t.

Also allow ml_dtypes as a local wheel override.

hawkinsp committed:
Move definition of pjrt_c_api_gpu_plugin.so into jax_plugins/{cuda,rocm}

This is simpler and should work better for a bazel submodule.

Google-ML-Automation committed:
Merge pull request #28750 from hawkinsp:requ

PiperOrigin-RevId: 758780857

Google-ML-Automation committed:
Merge pull request #28747 from jakevdp:canon-dtype-jax-array

PiperOrigin-RevId: 758785125

Google-ML-Automation committed:
Merge pull request #28751 from hawkinsp:plugins

PiperOrigin-RevId: 758801504

Google-ML-Automation committed:
Merge pull request #28748 from dfm:scan-lin

PiperOrigin-RevId: 758806795

hawkinsp committed:
Simplify the Bazel logic used to add dependencies to tests.

A lot of this logic was confusing phrased as conditions over both CPU
and GPU build flags. But we can decompose it:
* dependencies we add for CPU tests, and
* additional dependencies we add for GPU tests.

While we are here, also add the necessary pypi dependency for TPU tests.

pschuh committed:
Add use_raw_buffers which allows switching the implementation to
hold references to raw buffers instead of PjRtBuffers.

This fixes an issue where the buffers can be deleted before
the transfer is complete, but introduces another problem where
if they are donated it will now silently read from donated arrays.

Once the underlying runtime exposes usage holds properly, this
new codepath should take a usage hold and the old pjrtbuffer
path should be removed.

PiperOrigin-RevId: 758819621

Google-ML-Automation committed:
Merge pull request #28753 from hawkinsp:plugins

PiperOrigin-RevId: 758833461

hawkinsp committed:
Reenable CUDA version checks from Python.

These had been accidentally broken at some point in the plugin
switchover..

mattjj committed:
improve error message with when custom_vjp bwd rule produces wrong shape/dtype

yashk2810 committed:
Add numpy and absl/testing dep to custom_api_test. Fixes https://github.com/jax-ml/jax/actions/runs/15031061909/job/42243435305

PiperOrigin-RevId: 758898284

hyeontaek committed:
[JAX] Fix unhashable slice in api_test

`slice` is not hashable before Python 3.12. This change mitigates it by
converting it into a hash value.

PiperOrigin-RevId: 758905560

Google-ML-Automation committed:
Merge pull request #28755 from hawkinsp:plugins

PiperOrigin-RevId: 758915292

hawkinsp committed:
Fix CI build failure on Mac.

We must not depend on the nvidia_nvshmem_cu12 pip package directly since it does not exist on Windows and Mac platforms.

PiperOrigin-RevId: 758917499

superbobry committed:
[pallas] Do not emit verbose lowering errors by default

The errors are too verbose and mostly not very useful.

PiperOrigin-RevId: 759025165

bartchr808 committed:
#sdy Properly handle token types in JAX and `ManualComputationOp`.

We weren't handling them correctly meaning you couldn't use a `shard_map`/`ManualComputationOp` which has callbacks inside.

PiperOrigin-RevId: 759072597

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/ab7cea20271d8a24a7309e09fc5af486dde8e155.

PiperOrigin-RevId: 759095567

bartchr808 committed:
#sdy Fix incorrect sharding on a token during a callback.

The "add a token" part of the `callback` primitive's MLIR lowering was incorrectly adding a ranked sharding by using the sharding of a ranked tensor. So instead create an unranked sharding explicitly

PiperOrigin-RevId: 759135477

Google-ML-Automation committed:
Merge pull request #28757 from mattjj:custom-vjp-aval-mismatch-extra

PiperOrigin-RevId: 759203972

hawkinsp committed:
Workaround a crash on aarch64 due to a NumPy bug.

See: https://github.com/numpy/numpy/issues/28843

Google-ML-Automation committed:
Merge pull request #28771 from hawkinsp:numpy2

PiperOrigin-RevId: 759221096

Google-ML-Automation committed:
Merge pull request #28317 from yhtang:pr-add-k8s-docs

PiperOrigin-RevId: 759252455

jakevdp committed:
[pre-commit] update mypy to v1.15.0

Google-ML-Automation committed:
Merge pull request #28721 from jakevdp:fix-typestubs

PiperOrigin-RevId: 759294851

jakevdp committed:
[pre-commit] bump pre-commit version to v5.0.0

jakevdp committed:
[pre-commit] bump ruff to v0.11.9

Google-ML-Automation committed:
Merge pull request #28779 from jakevdp:pre-commit-version-bump

PiperOrigin-RevId: 759301396

Google-ML-Automation committed:
Merge pull request #28781 from mattjj:hijax

PiperOrigin-RevId: 759336328

mattjj committed:
[hijax] type-changing boxes with pytree contents

Google-ML-Automation committed:
[JAX] Add ticks around input to clearify in the error message. Currently it looks like this.

```
ValueError: Pytree for `in_specs` and inputs do not match. There are 1 mismatches, including:
    * `in_specs` is a tuple of length 1 but inputs is a tuple of length 4, so the lengths do not match

```

PiperOrigin-RevId: 759499528

vfdev-5 committed:
Added dict_dict_merge/split_keys_entry_added suppression to 3.14 TSAN CI job

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/5fee96f09a42daa80283dde9fb7090ba90d9d07a.

PiperOrigin-RevId: 759564260

Google-ML-Automation committed:
Merge pull request #28785 from vfdev-5:race-suppression-314-dict_dict_merge

PiperOrigin-RevId: 759579563

hawkinsp committed:
Add OMP_NUM_THREADS workaround to more aarch64 CI configuations.

Google-ML-Automation committed:
Merge pull request #28792 from hawkinsp:aarch64

PiperOrigin-RevId: 759602792

superbobry committed:
[mosaic_gpu] Added support for attaching source information to the PTX

The implementation currently forces O=0 due to a suspected bug in the NVPTX
backend.

To get source information

* Set MOSAIC_GPU_LINE_INFO=1
* Run with --jax_include_full_tracebacks_in_locations=true

PiperOrigin-RevId: 759608368

belitskiy committed:
Strip leading zeros from ML_WHEEL_GIT_HASH.

They end up being stripped by setuptools, which leads to a mismatch
between expected and actual wheel names, which is fatal, as Bazel is
expecting only a particular name, not to mention other issues.

https://peps.python.org/pep-0440/

PiperOrigin-RevId: 759610274

bchetioui committed:
[Pallas/Mosaic GPU] Fix `get_p` lowering to handle `RefUnion`s correctly.

In the case of transformed refs, it's possible that a transform ends up
changing the dtype of the reference considered. This typically happens when
extracting an aliased ref out of a `RefUnion`, but it could also happen if
we were to handle `RefBitcast`s. As a result, querying the dtype of refs by
querying their `aval` is not a safe operation.

PiperOrigin-RevId: 759624846

apaszke committed:
[Mosaic GPU] Skip tests that need stdout capture when using pytest

The prints all go to the stdout captured by pytest instead of being
intercepted by `jtu`.

PiperOrigin-RevId: 759656514

mattjj committed:
quick fix for internal test broken by #28781

PiperOrigin-RevId: 759658054

Google-ML-Automation committed:
Merge pull request #28778 from jakevdp:mypy-version-bump

PiperOrigin-RevId: 759668066

Google-ML-Automation committed:
Merge pull request #28780 from jakevdp:ruff-version-bump

PiperOrigin-RevId: 759678379

hawkinsp committed:
Fix test failures in GPU CI.

Google-ML-Automation committed:
Merge pull request #28802 from hawkinsp:ci

PiperOrigin-RevId: 759721174

bythew3i committed:
[Mosaic TPU][NFC] Consolidate `getIntConst`.

PiperOrigin-RevId: 759728429

junwhanahn committed:
Keep the serialized version of `BufferAssignmentProto` in `CompiledMemoryStats` to reduce its overheads

Most users of `CompiledMemoryStats` do not use this field. So it is cheaper in terms of both CPU and RAM to keep it as a serialized string rather than a proto. If this continues to become a problem, we can consider inventing a separate executable API for buffer assignment.

PiperOrigin-RevId: 759735092

brills committed:
Added a test case to guard JAX ad, jax2tf in jax.lax.scan.

PiperOrigin-RevId: 759808188

Google-ML-Automation committed:
Merge pull request #28804 from mattjj:hijax

PiperOrigin-RevId: 759960246

hawkinsp committed:
Increase some test shardings to work around CI timeouts.

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/582b15b6146c3c20ecc88cc2fc7dfddbe3e63dc0.

PiperOrigin-RevId: 759975157

yliu120 committed:
shard map pbroadcast

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/d4ff3a507f46f8102065b598456a99dc17758394.

PiperOrigin-RevId: 760260591

Google-ML-Automation committed:
Merge pull request #28808 from hawkinsp:ci

PiperOrigin-RevId: 760263370

superbobry committed:
[pallas] Generalized `empty_like` to accept a pytree of shapes/dtypes

PiperOrigin-RevId: 760529300

superbobry committed:
[pallas:mosaic_gpu] Added support for unrolling to `lax.fori_loop` lowering

We currently require that `unroll` divides `length` for simplicity. This
restriction can be lifted later if/when necessary.

PiperOrigin-RevId: 760540276

apaszke committed:
[Mosaic GPU] Add support for fp8 types in WGMMA

PiperOrigin-RevId: 760551961

apaszke committed:
[Pallas:MGPU] Fix a test that assumes x32 mode by specifying dtype explicitly

PiperOrigin-RevId: 760576351

apaszke committed:
[Mosaic GPU] Ignore singleton tiling dims when constructing nested shape

They don't contribute anything, but they can trigger some NotImplemented errors
down the line.

PiperOrigin-RevId: 760579238

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/821e786cf9e478c8d866610203a80acaa70539b9.

PiperOrigin-RevId: 760580406

bartchr808 committed:
#sdy fix `shard_map` lowering if there is only one device.

This was wrong when the shmap contains callbacks, which cause tokens to be created. I was calling `jaxpr_subcomp` incorrectly.

PiperOrigin-RevId: 760588882

apaszke committed:
Add missing test skips and improve compatibility with older jaxlib versions

PiperOrigin-RevId: 760589385

apaszke committed:
[Mosaic GPU] A handful of minor bug fixes

PiperOrigin-RevId: 760601235

yashk2810 committed:
Add version guards on unreduced lowering to shardy

PiperOrigin-RevId: 760603883

hawkinsp committed:
Disable cusolver version check.

This appears to be failing for some reason, and it's safest just to disable it for now.

hawkinsp committed:
Bump NumPy version to 2.2.6 for 3.13+ Python versions.

Remove workaround for NumPy 2.2 bug on aarch64.

Also add a linux system constraint to GPU wheels.

Rifur13 committed:
[Mosaic GPU] Adding an optional causal masking to the manual pipelining example.
Testing shows the runtime is about half the flops with causal masking vs without.

Tests pass if we revert to cuda 12.0 with `--//third_party/gpus/cuda:by_exception_only_cuda_version_override=12_0`

PiperOrigin-RevId: 760616647

apaszke committed:
Fix the CAS loop in semaphore_wait lowering

The loop incorrectly assumed that we're waiting for the semaphore
value to be equal to the wait value, so that we can reset it to 0.
This, however, is not how semaphores work. The CAS loop should wait
until the value is _at least_ the wait value, and should update its
expectation at every step in case the swap failed.

The attached test deadlocks with the original implementation, but
works fine with the new one.

PiperOrigin-RevId: 760619094

dfm committed:
Fix scipy nightly CI failures caused by toeplitz.

Google-ML-Automation committed:
Merge pull request #28828 from hawkinsp:vercheck

PiperOrigin-RevId: 760633986

Google-ML-Automation committed:
Merge pull request #28829 from hawkinsp:req

PiperOrigin-RevId: 760636732

MichaelHudgins committed:
[CI] Rollback move to gcloud cp for tpu job

PiperOrigin-RevId: 760640560

Google-ML-Automation committed:
Merge pull request #28812 from yliu120:yl/debug_pbroadcast

PiperOrigin-RevId: 760640954

Google-ML-Automation committed:
Merge pull request #28830 from dfm:toeplitz-nightly

PiperOrigin-RevId: 760694785

justinjfu committed:
[Pallas] Fix 1D Iota

PiperOrigin-RevId: 760705572

rickeylev committed:
cleanup: load python rules from rules_python

belitskiy committed:
Strip leading zeros local version in all the necessary places, for wheel name.

PiperOrigin-RevId: 760830953

bixia1 committed:
Propagate use_shardy_partitioner to XlaCallModule op.

PiperOrigin-RevId: 760876735

jaswanth- committed:
[Mosaic] Add support for currently unsupported reshapes for 32-bit datatypes with native tiling and adds tests for those cases. The cases supported are (k % 128 == 0 in the below):
- (q, m, n, k) -> (q, m, n * k)
- (p, q, m, n, k) -> (p, q * m * n * k)
- (q, m, n, k) -> (q, m, 1, n * k) (in 2 steps, first to n*k then add unit dim)
- (q, m, n, k) -> (q * m, n * k)
- (q * m, n, k) -> (q, m, n * k)
- (q * m, n * k) -> (q, m, n, k)
- (q, m, n * k) -> (q * m, n, k)

PiperOrigin-RevId: 760904758

Google-ML-Automation committed:
Automated Code Change

PiperOrigin-RevId: 760909620

WindQAQ committed:
[Mosaic] Use native bf16 ops for tanh, exp and log on TPUv6+.

Replace `needs_cast` condition during canonicalization with `need_elementwise_canonicalization`.

PiperOrigin-RevId: 760934277

Google-ML-Automation committed:
Automated Code Change

PiperOrigin-RevId: 760956866

superbobry committed:
[pallas] Pulled `runtime_assert_enabled` from `pltpu` to `pl`

PiperOrigin-RevId: 760983479

superbobry committed:
[pallas:mosaic_gpu] Added `plgpu.nd_loop`

This is a generalization of `lax.fori_loop` which partitions the flat iteration
space across the given axes, and is useful for writing persistent kernels.

PiperOrigin-RevId: 761007326

dfm committed:
Update scipy.signal.welch tests to be compatible with upstream dev version.

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/4a914a285377e90c464a0b7fad9b5cbcfeeb27a9.

PiperOrigin-RevId: 761022928

apaszke committed:
[Mosaic GPU] Implement reshapes from and to refs with empty shapes

PiperOrigin-RevId: 761038562

hawkinsp committed:
Add Python 3.13 support to traceback code under PLATFORM_GOOGLE.

This build options uses CPython internals, which changed under Python 3.13.

PiperOrigin-RevId: 761042927

apaszke committed:
Only run relevant GPU tests in the optional GPU CI

This aims to prepare those tests to be used as a presubmit.
We can later add a more complete nightly configuration for testing.

PiperOrigin-RevId: 761044847

MichaelHudgins committed:
[CI] Increase CUDA pytest timeout.

PiperOrigin-RevId: 761047639

superbobry committed:
[mosaic_gpu] Removed `uniform=` from `async_copy`

It is redundant in the presence of `predicate=`.

PiperOrigin-RevId: 761085680

superbobry committed:
[pallas:mosaic_gpu] Removed the `GPU*` prefix from Mosaic GPU-specific types

These APIs are always used qualified, e.g. `plgpu.GPUCompilerParams`, so the
prefix is redundant.

PiperOrigin-RevId: 761088896

danielsuo committed:
Rename backend.compile to backend.compile_and_load.

Part of a larger refactor. Today, `compile` returns a loaded executable i.e., fuses the compile and load functions. Eventually, `compile` should return an unloaded executable and `load` should return a loaded exectuable; the default jit path will still return a loaded executable.

PiperOrigin-RevId: 761098001

Google-ML-Automation committed:
Merge pull request #28848 from dfm:fix-welch-ci

PiperOrigin-RevId: 761101581

hawkinsp committed:
Add a TSAN suppression for https://github.com/python/cpython/issues/132214 under Python 3.13.

Rifur13 committed:
[Mosaic GPU] Replace `core_map` + `run_state` with `plgpu.kernel` for simpler code. Slightly more efficient because we don’t initialize the outputs now.

PiperOrigin-RevId: 761113870

superbobry committed:
[pallas:mosaic_gpu] Use `MemorySpace` aliases

This just makes the corresponding conditions a bit easier to read.

PiperOrigin-RevId: 761137840

dfm committed:
Include Pallas TPU random ops in JAX wheel.

Since the `pallas/tpu/ops/random` directory was missing an `__init__.py` file, it was inadvertently excluded from the released JAX distribution. I don't see any reason why this submodule shouldn't be included so let's fix that!

To deal with the fact that they weren't included in the distribution, we were also monkey patching these files into the wheel when testing, but that's no longer needed.

PiperOrigin-RevId: 761138525

danielsuo committed:
Migrate users of backend.compile to backend.compile_and_load.

PiperOrigin-RevId: 761150621

Google-ML-Automation committed:
Merge pull request #28856 from hawkinsp:tsan

PiperOrigin-RevId: 761152994

sashabu committed:
[Mosaic] Tweak `tpu.log` verification on SC.

PiperOrigin-RevId: 761182341

Google-ML-Automation committed:
Merge pull request #28838 from rickeylev:load.py.rules

PiperOrigin-RevId: 761191493

hawkinsp committed:
Remove tsan suppression from python/cpython#129748

This bug is marked as fixed upstream.

yashk2810 committed:
Remove pspec -> names conversion during `shard_map_p.bind` and instead preserve partition specs everywhere internally.

**This is because spec -> names canonicalization gets rid of unreduced axes present on PartitionSpecs and we want to preserve that**. We can thread 2 new parameters called `in_unreduced` and `out_unreduced` and keep `in_names`, `out_names` but that doesn't buy us anything except for more lines added and complexity :)

It's better to just use pspecs everywhere. It's a net reduction in lines of code too!

PiperOrigin-RevId: 761196531

sizhit2 committed:
[JAX] Enable tfrt gpu in jax multi platform test

PiperOrigin-RevId: 761204561

Commit:
[chex] remove stale try/except import

chex now requires jax>=0.4.27, so the previous backward-compatibility is no longer necessary.

PiperOrigin-RevId: 761212887

bythew3i committed:
[ragged-paged-attn] Apply kv mask to filter out NaNs

Expected small regression as we insert kv masking logic which needs to unpack/pack.

PiperOrigin-RevId: 761217303

tlongeri committed:
[Mosaic:TPU] Enforce that tpu.dynamic_gather operands and result have the same shape

PiperOrigin-RevId: 761230458

Google-ML-Automation committed:
Make sure tests with `--build_jaxlib=false` depend on NVIDIA CUDA wheels hermetically.

PiperOrigin-RevId: 761231648

hawkinsp committed:
Reverts 06448864abd6e8187e5b4d9b1ff08ab14fe3b8e0

PiperOrigin-RevId: 761237485

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/ea0d903efdff87b261e1d68c59011138ee13a9ac.

PiperOrigin-RevId: 761250461

Google-ML-Automation committed:
[JAX] Fix float typo in a code example in the sharded-computation doc.

PiperOrigin-RevId: 761268857

Google-ML-Automation committed:
Merge pull request #28863 from hawkinsp:tsan2

PiperOrigin-RevId: 761283866

hawkinsp committed:
Do not construct an error string for FoldingError.

This just wastes time since the error will never be read.

PiperOrigin-RevId: 761285843

hawkinsp committed:
Reland the limit on the number of OpenBLAS threads.

This was previously removed in https://github.com/jax-ml/jax/commit/18ff6caa4f767701dd7cca3a1333d9b99465e045, and that promptly broke our CI again. I am guessing the problem is actually too few threads, not a NumPy deadlock as I originally guessed.

Google-ML-Automation committed:
Merge pull request #28871 from hawkinsp:ci

PiperOrigin-RevId: 761306262

yashk2810 committed:
Support passing PartitionSpecs to ShapeDtypeStruct when there is a mesh in context.

PiperOrigin-RevId: 761322712

apaszke committed:
Add missing test skips (for too old jaxlib) + bump minimum libtpu version

PiperOrigin-RevId: 761429155

apaszke committed:
Disable the newly added tfrt targets that never worked

PiperOrigin-RevId: 761442340

apaszke committed:
Disable tests for the paged attention kernel

I ran the TPU race checker on it and it did report a number of races
that were uncovered by recent Mosaic compiler changes.

PiperOrigin-RevId: 761447182

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/47a47140b5f7a33d86901316d9c569de179bae08.

PiperOrigin-RevId: 761478240

superbobry committed:
[pallas:mosaic_gpu] Removed `GPU*` aliases

PiperOrigin-RevId: 761482778

bchetioui committed:
[Mosaic GPU] Support `s8xs8->s32` WGMMA.

PiperOrigin-RevId: 761489756

bchetioui committed:
[Pallas/Mosaic GPU] Loosen tiling requirements for `get` and `swap`.

We now allow arbitrary 2D tilings where the minor dimension fits the associated
swizzle.

PiperOrigin-RevId: 761490845

dfm committed:
Skip generating source links for re-exported numpy functions in docs.

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/c74abfc3ecfdc31901acca65efa29ffda3ed84cc.

PiperOrigin-RevId: 761521044

dfm committed:
Try using uv for installing packages on Read the Docs.

hawkinsp committed:
Prepare for jax release v0.6.1

nitins17 committed:
Skip oldest supported numpy presubmit on release branches

PiperOrigin-RevId: 761534198

Google-ML-Automation committed:
Merge pull request #28887 from dfm:uv-on-rtds

PiperOrigin-RevId: 761545482

Google-ML-Automation committed:
Merge pull request #28886 from dfm:np-doc-links

PiperOrigin-RevId: 761552874

olupton committed:
Revert "jax-cuda12-plugin: require nvidia-cublas-cu12<12.9"

Google-ML-Automation committed:
Introduce the flag `add_pypi_cuda_wheel_deps` that controls if the tests depend on NVIDIA CUDA wheels hermetically.

The flag is enabled by default.

To disable the dependency, pass `add_pypi_cuda_wheel_deps=False` in the Bazel options.

PiperOrigin-RevId: 761568590

rdyro committed:
Adjust triton dialect lowering rounding mode to allow upcasting fp8 types

Fix: https://github.com/jax-ml/jax/issues/28416
PiperOrigin-RevId: 761577943

WindQAQ committed:
[Mosaic] Add faster implementation for s8->bf16 and s4->bf16 on TPUv6+.

PiperOrigin-RevId: 761578503

belitskiy committed:
Show exactly what the copy paths/patterns are when copying wheels.

This will make it easier to track down unexpected path mismatches in the future.

PiperOrigin-RevId: 761584888

jenriver committed:
Use the default python logging instead of absl log.

Update use cases in export / compiler as well as the documentation.

Google-ML-Automation committed:
Merge pull request #28897 from olupton:revert-28513-skip-cublas-12.9

PiperOrigin-RevId: 761587875

Google-ML-Automation committed:
[ragged-paged-attn] Use select for initialization in flash attention.

PiperOrigin-RevId: 761612390

hawkinsp committed:
Merge branch 'release/0.6.1'

hawkinsp committed:
Update version numbers after 0.6.1 release.

tlongeri committed:
[Mosaic:TPU][Relayout] Support minor to 2nd minor implicit dimension for unpacked types and native tiling on TPUv5

PiperOrigin-RevId: 761676578

Commit:
Move _src/stages.py to its own build target

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

This refactor required moving the definitions of a few private utilities from pjit and pxla, because these files are part of the larger jax build target.

PiperOrigin-RevId: 761689391

Google-ML-Automation committed:
Merge pull request #28904 from hawkinsp:postrelease

PiperOrigin-RevId: 761690584

tlongeri committed:
[Mosaic:TPU][Relayout] Remove minor implicit dimension for 32-bit native tiling

PiperOrigin-RevId: 761692972

levskaya committed:
Expose tree prefix broadcasting as a public API in tree utils.

yashk2810 committed:
Allow eval_shape to propagate shardings if the aval has shardings in full explicit mode

PiperOrigin-RevId: 761708753

Commit:
Move _src/interpreters/ad.py to its own BUILD rule.

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

This required moving some internal utilities out of dispatch.py, which is part of the main JAX build rule. I chose api_util.py because they seem to fit there.

PiperOrigin-RevId: 761722054

Google-ML-Automation committed:
Add initial support aligned jnp.swapaxes on major/minor dims

Next steps:
  - non-tile aligned
  - Clean up fn and utilize it for general changeTiling

PiperOrigin-RevId: 761731600

yashk2810 committed:
remove jaxlib_extension_version, ifrt_version and jaxlib.__version_info__ guards after 0.6.1 release.

PiperOrigin-RevId: 761737523

yashk2810 committed:
Add out_sharding to the function returned by `jax.nn.initializers.he_normal` and other APIs implementing the `Initializer` protocol. Currently it takes `key, shape, dtype` and now we added an optional out_sharding parameter to it.

PiperOrigin-RevId: 761742909

Commit:
Add visibility hook for //jax:stages

PiperOrigin-RevId: 761745409

Google-ML-Automation committed:
Merge pull request #28909 from levskaya:tree_broadcast

PiperOrigin-RevId: 761758158

JyotinderSingh committed:
Fix `jax.grad()` documentation in quickstart.md

The documentation wrongly mentions that the `sum_logistic` function was jitted in the preceding example - which is not true.

Fixed the phrasing to be more accurate.

mattjj committed:
fix custom_vjp optimize_remat=True with collectives

Google-ML-Automation committed:
Merge pull request #28923 from mattjj:rahul-fix3

PiperOrigin-RevId: 761806312

apaszke committed:
Fix skip condition in Pallas Triton tests

It should not run in TPU configs at all.

PiperOrigin-RevId: 761865943

Google-ML-Automation committed:
[Pallas] Simulate multiple cores per device when interpreting kernels on CPU.

The `TPUInterpreterParams` are extended with a field `num_cores_per_device` that specifies how many cores should be simulated per (TPU) device.
Per-core devices are mapped along grid dimensions with _parallel_ dimension semantics.
(I.e. the body of the grid loop is considered to execute on the next device-local core if the multi-dimensional loop index has changed along a _parallel_ grid dimension from one loop iteration to the next.)

Each per-device core is identified by a `local_core_id`. Globally, each core is identified by a `global_core_id` (= `device_id` * `num_cores_per_device` + `local_core_id`).
There is one vector clock per core (previously, one per device), and each vector clock has as many entries as there are cores in total (= `num_device` * `num_cores_per_device`).

PiperOrigin-RevId: 761872013

bchetioui committed:
[Pallas/Mosaic GPU] Add a Pallas:MGPU implementation of `ragged_dot`.

PiperOrigin-RevId: 761875827

apaszke committed:
[Pallas:MGPU] Make semaphores compatible with plgpu.kernel and the profiler

The previous code was overly specific to pl.pallas_call and did not work with
plgpu.kernel at all. Now, semaphores can be allocated using run_scoped, which also
has the interesting side effect of the allocations being collective within
each program. For a persistent kernel that means that a program/block can communicate
with programs/blocks on other devices that have the same program ID, but it is
currently impossible to e.g. synchronize all programs in a grid. We don't have a
use case for it now, so we can add it later.

PiperOrigin-RevId: 761885876

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/f040466fc5fa5052b1f4e89fb2b166bb2c9d656a.

PiperOrigin-RevId: 761914597

apaszke committed:
[Pallas:MGPU] Add a first prototype of an all_gather collective matmul kernel

It's not very optimized at the moment and is unlikely to outperform the baseline
of raw all_gather + matmul, but it computes the right numbers. We are already aware
of a few places that could be optimized and we'll start rolling them out soon.

PiperOrigin-RevId: 761939624

apaszke committed:
[Pallas:MGPU] Disable the GPU distributed test when the platform allocator is used

And, if we're only running this test, try to override the flag.

PiperOrigin-RevId: 761950583

jburnim committed:
Prevent tests of TPU interpret mode from being run in parallel.

Google-ML-Automation committed:
Merge pull request #28922 from JyotinderSingh:patch-1

PiperOrigin-RevId: 761960261

apaszke committed:
Bump the oldest libtpu version again

The previous bump was wrong because apparently there was a 3 day window
in which we haven't been publishing nightlies. This time, I checked that
there is a package released on that day.

PiperOrigin-RevId: 761963834

Google-ML-Automation committed:
Merge pull request #28935 from jburnim:jburnim_interpret_test

PiperOrigin-RevId: 761974061

yashk2810 committed:
Fix `test_make_array_from_single_device_arrays` in random_test.py for TPU 7x. This is because `create_mesh` return a different order of devices while `arrays` are being created with `jax.devices()` order.

PiperOrigin-RevId: 761975228

superbobry committed:
[pallas:mosaic] Enabled `scan` lowering rule for all kernel types

PiperOrigin-RevId: 761979336

apaszke committed:
[Pallas:MGPU] (Slightly) Clean up the collective matmul test by using explicit sharding

Not a huge difference, but it's a bit nicer.

PiperOrigin-RevId: 761984854

dfm committed:
Reduce input ranges tested for hyp2f1.

The `z` input to this function only supports input in the range 0 <= z < 1, so most of the tested samples were explicitly `inf`. The other parameters can take arbitrary values, but the dynamic range required explodes for values larger than ~1, so the old tests were quite numerically unstable. We could test some specific carefully chosen larger values, but I don't think this brings much benefit.

PiperOrigin-RevId: 762000716

apaszke committed:
[Mosaic GPU] Don't require MOSAIC_GPU_NVSHMEM_SO_PATH to be set

PiperOrigin-RevId: 762008659

hawkinsp committed:
Update the trove classifiers for JAX packages.

* Don't tag packages as alpha.
* Tag free-threading as supported.
* Update some python version lists.

Google-ML-Automation committed:
Merge pull request #28945 from hawkinsp:trove

PiperOrigin-RevId: 762016836

jakevdp committed:
jax.random: thread mode parameter through categorical and choice

yashk2810 committed:
Error out if wsc(x, P()) is called in a Explicit mesh context

PiperOrigin-RevId: 762021159

justinjfu committed:
[Mosaic GPU] Add support for loops, debug_print, and unary ops to Warp semantics.

PiperOrigin-RevId: 762036132

rdyro committed:
Add support for non-power-of-2 head size in flash attention
Introduce checks on sequences being divisible by block sizes to address https://github.com/jax-ml/jax/issues/27224

PiperOrigin-RevId: 762051831

Google-ML-Automation committed:
Merge pull request #28946 from jakevdp:random-mode

PiperOrigin-RevId: 762092464

Google-ML-Automation committed:
Merge pull request #28899 from jenriver:absl_logging_fix

PiperOrigin-RevId: 762099728

sizhit2 committed:
Remove block_until_ready from testAutodiffCache

PiperOrigin-RevId: 762116819

nitins17 committed:
Add CUDA pytest job strategy to test CUDA packages downloaded from PyPI

Continuous and Nightly/Release workflows will now run the CUDA 12.8 test runs by using the Nvidia CUDA packages from PyPI instead of those on the system.

PiperOrigin-RevId: 762125356

Google-ML-Automation committed:
Add pypi NVIDIA NVCC wheel dependency to the tests that run with pre-built wheels.

NCCC wheel is used in `[with-cuda]` requirements as stated [here](https://github.com/jax-ml/jax/blob/0dc70b93f2e13fae5b097837760bd621e746dae7/jax_plugins/cuda/plugin_setup.py#L58).

PiperOrigin-RevId: 762126511

jkr26 committed:
Raises an explicit error in reshard.

Hit this while working with sharding in types -- passing a sharding that had an empty mesh. (I think this was in a test). This failed trying to acces with `with_spec` attribute on None -- so just catching this case early.

PiperOrigin-RevId: 762135310

jburnim committed:
Use additional semaphores to avoid data races in TPU paged_attention_kernel.

Also prevents an out-of-bounds read of SMEM.  And re-enables tests for the TPU paged_attention_kernel.

@apaszke confirmed the presence of data races using the race detector in the new TPU interpret mode.  With the additional semaphores, the race detector no longer detects any races in the this kernel and I no longer see any test failures in 20+ test runs on a TPU.

Details on the data races:

 - In each iteration, the kernel:
   (a) Starts copying data for `k` and `v` for the next iteration.
   (b) Waits for the copy of `k` for the current iteration to finish.
   (c) Waits for the copy of `v` for the current iteration to finish.

 - It is possible for these copies to happen out of order -- that is:
   (a) The copies for the next iteration can finish before the copies
       for the current iteration.
   (b) And the copies for `v` for the current iteration can finish
       before the copies for `k` for the current iteration.

 - If the same DMA semaphore is used for everything, then out-of-order
   copies can lead to:
   (a) `k = async_copy_k.wait_and_get_loaded()` returns but the data
       isn't all available because the underlying semaphore was
       signaled by the completion of copies of `v` for the current
       iteration or copies of `k` or `v` for the next iteration.
   (a) `v = async_copy_v.wait_and_get_loaded()` returns but the data
       isn't all available because the underlying semaphore was
       signaled by the completion of copies of `k` or `v` for the
       next iteration.

PiperOrigin-RevId: 762136079

yashk2810 committed:
Adds explicit-axis handling to shard_map batching rule.

Without this handling, in explicit sharding mode vmap of a function with an internal shmap can introduce unnecessary replication.

PiperOrigin-RevId: 762175189

ZixuanJiang committed:
Introduce `get_axis_sizes` API for `HloSharding` in JAX.

PiperOrigin-RevId: 762213931

ZixuanJiang committed:
Implement `_to_sdy_sharding` for GSPMDSharding.

PiperOrigin-RevId: 762220186

yashk2810 committed:
Fix psum_invariant transpose rule where we were binding `pvary` with `ad_util.Zero` which lead to errors like this: `TypeError: Argument 'Zero(float32[1,1,512])' of type '<class 'jax._src.ad_util.Zero'>' is not a valid JAX type`

PiperOrigin-RevId: 762264425

dimitar-asenov committed:
[Mosaic GPU] Properly handle single-element outputs in inline assembly.

This issue was discovered when enabling the ragged dot example kernel to run using warpgroup semantics.

The new test requires `-UNDEBUG`.

PiperOrigin-RevId: 762365352

apaszke committed:
[Mosaic GPU] Add support for async copies to peer devices

PiperOrigin-RevId: 762370447

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/c361fc2992e8d674636e7870992e95658b1be792.

PiperOrigin-RevId: 762376295

apaszke committed:
[Mosaic GPU] Add checks for argument shapes and types

Apparently we never checked it and it's been quite easy to
get this wrong.

PiperOrigin-RevId: 762394139

dimitar-asenov committed:
[Mosaic GPU] Pass the right number of immediate values in the wgmma inline asm.

Before this change, in the int32 case, we pass two extra immediate args compared to the number of parameters in the ASM string. Running tests with `-UNDEBUG` detects the error.

I think this likely broke with the special-casing of `int32` in cl/761489756.

I've added an assert that should prevent mismatches in the future.

PiperOrigin-RevId: 762394530

tomnatan30 committed:
Fix sharding rule in jax test

`ValueError: Sharding rule has 1 operands, but the operation has 2 operands`

PiperOrigin-RevId: 762412744

Commit:
Move _src/custom_transpose.py into its own BUILD rule

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

PiperOrigin-RevId: 762414305

dfm committed:
pytest: use importlib mode by default.

This is an attempt to re-land https://github.com/jax-ml/jax/pull/28650, fixing build failures.

**Motivation**

https://github.com/jax-ml/jax/pull/28650 was motivated by recent changes in setuptools, which caused test failures with editable installs, but it also exposes a potentially larger issue with our approach to testing with pytest. As far as I understand it, the default pytest behavior is to prepend the working directory to `sys.path`, meaning that `jax` is imported from the source directory, rather than the installed version. Switching to the `importlib` import mode means that we correctly test against the installed version of `jax`, which seems like what we typically want to do.

The catch is that then we need to explicitly package any test utilities into the distribution. We don't currently package test-specific utilities like `internal_test_util` with JAX, but these utilities were still available to tests since they live within the `jax` source tree. This breaks when using `importlib` import mode, and a non-editable install of JAX.

**Solutions**

The approach that I've taken here is to explicitly package everything needed by the tests into the `jax` distribution. This means that we can correctly test against the _installed_ version of JAX when using pytest.

This solution isn't ideal because it means that we're distributing `jax` submodules that aren't actually required except when running the test suite, but this seems like a small price to pay to me.

**Alternatives**

One different approach that we could take would be to only support using pytest with _editable_ installs of JAX. This would work because the required files would still be discoverable in an editable install because they live within the source tree. In fact, before this change, most of our CI jobs actually did install an editable distribution (which is why the failures in https://github.com/jax-ml/jax/pull/28650 weren't caught in pre-submit!). The problem with this approach is that we're not actually testing JAX as it is used when installed from a distribution, and it wouldn't catch things like missing submodules. I think it's much better to test against the installed distribution!

A more extreme approach would be to switch JAX to a `src/jax` and `src/jaxlib` layout (i.e. moving `jax` and `jaxlib` out of the root directory) as recommended by the Python packaging docs. Unfortunately this would be complicated with the way JAX is distributed internally at Google, so I think that's probably a non-starter.

PiperOrigin-RevId: 762419160

superbobry committed:
[pallas] Slightly revamped how checkify is exposed in Pallas

* We now re-export a restricted version of `debug_check` under `pl`.
  Unlike the original, the `pl` version only allows a static message, i.e.
  string interpolation is not supported.
* Only debug checks are supported, which means that by default no checking
  is done -- `debug_check` is lowered to a noop.
* The context manager enabling debug checks is called `enable_debug_checks`.
  I would very much like to drop the `enable_` prefix, but without it the
  context manager reads too similar to `debug_check`.

PiperOrigin-RevId: 762433258

danielsuo committed:
[jaxlib] Add CompileOnlyPyClient to xla_client.

We have users of CompileOnlyPyClient that use `backend.compile` as we eventually intend it (i.e., return `ExecutableRef`, possibly `PyExecutable` eventually, instead of `PyLoadedExectuable`).

PiperOrigin-RevId: 762440439

MichaelHudgins committed:
[CI] Pin the ML Connect action to a specific sha

PiperOrigin-RevId: 762441171

Commit:
Move jax/_src/interpreters/batching.py into its own BUILD rule

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

Unfortunately this is not a clean build refactor, because batching depends on jax.lax, which in turn depends on batching. However, the problematic functions are only called within contexts where jax.lax is available for import.

We have a few options here:

1. Continue to bundle the batching.py source with the main build.
2. Build separately, but do the local import workaround in this CL (a pattern we use elsewhere).
3. Build this separately, but move some batching definitions into jax.lax for a more strict dependency graph. Or pass the `lax` namespace explicitly to the function at the call site.

I opted for (2) here because I judged the benefits of a refactored build to be worth the cost of localized impure dependencies, and the kind of refactoring in (3) would affect some downstream users.

PiperOrigin-RevId: 762447323

jburnim committed:
Add TODO to run TPU interpret mode tests in parallel.

PiperOrigin-RevId: 762456274

danielsuo committed:
Rename backend.compile to backend.compile_and_load.

Part of a larger refactor. Today, `compile` returns a loaded executable i.e., fuses the compile and load functions. Eventually, `compile` should return an unloaded executable and `load` should return a loaded exectuable; the default jit path will still return a loaded executable.

PiperOrigin-RevId: 762457830

matthiasdiener committed:
Fix Gotchas link

jakevdp committed:
[tree_util] raise more informative error when pytree equality check fails

justinjfu committed:
[Mosaic GPU] Add support for copy_gmem_to_smem in Warp semantics.

PiperOrigin-RevId: 762475094

Google-ML-Automation committed:
Merge pull request #28978 from matthiasdiener:patch-1

PiperOrigin-RevId: 762490961

justinjfu committed:
[Mosaic GPU] Add barrier transformation support to tcgen05_mma.

Also fix accumulator argument when it's dynamic.

PiperOrigin-RevId: 762509416

Commit:
Move jax/_src/custom_dce.py to its own BUILD rule

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

This was unblocked by moving batching & ad to their own rules in prior changes.

PiperOrigin-RevId: 762527517

pschuh committed:
Transfer library: poison outstanding buffer fetches upon connection failure.

PiperOrigin-RevId: 762546985

ZacCranko committed:
This is a change to patch some internal Google builds while we complete a refactor.

PiperOrigin-RevId: 762547026

Google-ML-Automation committed:
Merge pull request #28979 from jakevdp:pytree-err

PiperOrigin-RevId: 762562570

rdyro committed:
Simplify attention VJP definition

PiperOrigin-RevId: 762567722

Google-ML-Automation committed:
[ragged-paged-attn] Implement static kv cache quantization. (The scale of kv cache is a scalar float value)

PiperOrigin-RevId: 762576286

Commit:
Move jax/_src/tree.py to its own build rule

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, and leads to improved build and iteration times.

PiperOrigin-RevId: 762589488

dlwh committed:
Update index.rst

add marin to ecosystem

Commit:
Move jax/_src/sourcemap to its own build rule

Creating smaller build rules enforces better organized dependency graphs in the JAX project, helps pytype propagate annotations correctly, discourages private imports downstream, and leads to improved build and iteration times.

PiperOrigin-RevId: 762621491

jburnim committed:
Use block_until_ready to fix races in TPU interpret mode tests

PiperOrigin-RevId: 762621608

Google-ML-Automation committed:
Merge pull request #28992 from dlwh:marin-link

PiperOrigin-RevId: 762623893

mrguenther committed:
Update XlaCallModule so tests are compatible with DCE.

Move the logic verifying that `shape_assertion` custom calls have side effects to run before MLIR optimizations are applied instead of after. Any `shape_assertion` custom call violating this condition (i.e. declared as pure) is likely to be removed by dead-code elimination, making it undetectable after optimizations. (Until recently, the test passed because DCE wasn't correctly applied to `custom_call` ops.)

PiperOrigin-RevId: 762634648

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/bc63ed41661e939ced1b3d3bfd7d3083eaabd747.

PiperOrigin-RevId: 762804332

yashk2810 committed:
Enter into the right mesh context during shmap DCE

PiperOrigin-RevId: 762956847

Google-ML-Automation committed:
Update XLA dependency to use revision
http://github.com/openxla/xla/commit/2c35b7a03388e9ee81b3a71037c372595232ff84.

PiperOrigin-RevId: 763076180

Commit:
Merge branch 'patch-1' of https://github.com/johannahaffner/jax into patch-1
